### PR TITLE
Add Consent Extension

### PR DESCRIPTION
**Pull request by cacheung**
Add Consent Extension Implementation
To Do:  Readme, Test, Update example App.

Current UI for example app with tabs.

consent

## Description
Add Consent Extension

Quick notes for running example app
Steps:
1. Install flutter (only one time for the machine)
https://docs.flutter.dev/get-started/install/macos

2.  In terminal window: (enable flutter for this terminal)
export PATH="$PATH:/Users/UserName/flutter/bin"

(Have the simulator or emulator launched first)

3. Browse to the repo example folder
e.g cd  /Users/UserName/aepsdk_flutter/example 

4. flutter pub get

5. flutter run

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: plugins/flutter_aepconsent/lib/flutter_aepconsent.dart [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+class Consent {
+  static const MethodChannel _channel = 
+       const MethodChannel('flutter_aepconsent');
+
+  /// Gets the current AEPConsent extension version.
+  static Future<String> get extensionVersion =>
+      _channel.invokeMethod('extensionVersion').then((value) => value!);
+
+  /// Retrieves the current consent preferences stored in the Consent extension
+  static Future<Map<dynamic, dynamic>> get consents => _channel
```
Comment (cacheung) [inline plugins/flutter_aepconsent/lib/flutter_aepconsent.dart:25]: I am getting below error if I set to <Map<String, dynamic>> , using <Map<dynamic, dynamic>> seems the return comes back properly.
[VERBOSE-2:ui_dart_state.cc(209)] Unhandled Exception: type '_InternalLinkedHashMap<Object?, Object?>' is not a subtype of type 'Map<String, dynamic>?' in type cast
#0      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:177:41)
<asynchronous suspension>
#1      _MyAppState.getConsent (package:flutter_aepsdk_example/main.dart:231:16)
<asynchronous suspension>

File: example/android/app/src/main/java/com/adobe/marketing/mobile/flutter/flutter_aepsdk_example/MyApplication.java [hunk 2]
```diff
@@ -23,7 +24,8 @@ public void onCreate() {
             Lifecycle.registerExtension();
             Signal.registerExtension();
             Assurance.registerExtension();
-            MobileCore.start(o -> MobileCore.configureWithAppID("yourAppId"));
+            Consent.registerExtension();
+            MobileCore.start(o -> MobileCore.configureWithAppID(""));
```
Comment (addb) [inline example/android/app/src/main/java/com/adobe/marketing/mobile/flutter/flutter_aepsdk_example/MyApplication.java:28]: Let's have a constant for App ID, similar to our Android/iOS samples.

Comment (addb) [inline example/android/app/src/main/java/com/adobe/marketing/mobile/flutter/flutter_aepsdk_example/MyApplication.java:28]: Also what is the `o` parameter for MobileCore.start API. Lets replace with `_` or a descriptive name.

File: plugins/flutter_aepconsent/README.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+ > Note: It is required to initialize the SDK via native code inside your AppDelegate and MainApplication for iOS and Android respectively.
+
+As part of the initialization code, make sure that you set the SDK wrapper type to `Flutter` before you start the SDK.
+
+Refer to the [Initialization](https://github.com/adobe/aepsdk_flutter#initializing) section of the root README for more information about initializing the SDK.
+
+##### Importing the SDK:
+```dart
+import 'package:flutter_aepassurance/flutter_aepconsent.dart';
```
Comment (addb) [inline plugins/flutter_aepconsent/README.md:38]: typo `package:flutter_aepconsent/flutter_aepconsent.dart`

File: plugins/flutter_aepconsent/android/build.gradle [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+}
+
+apply plugin: 'com.android.library'
+
+android {
+    compileSdkVersion 28
+
+    defaultConfig {
+        minSdkVersion 16
```
Comment (emdobrin) [inline plugins/flutter_aepconsent/android/build.gradle:28]: min should be 19

File: plugins/flutter_aepconsent/ios/Classes/FlutterAEPConsentPlugin.m [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            binaryMessenger:[registrar messenger]];
+  FlutterAEPConsentPlugin* instance = [[FlutterAEPConsentPlugin alloc] init];
+  [registrar addMethodCallDelegate:instance channel:channel];
+}
+
+- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
+  if ([@"extensionVersion" isEqualToString:call.method]) {
+      result([AEPMobileEdgeConsent extensionVersion]);
+  } else if ([@"getConsents" isEqualToString:call.method]) {
+    [self handleGetConsent:call result:result];
```
Comment (emdobrin) [inline plugins/flutter_aepconsent/ios/Classes/FlutterAEPConsentPlugin.m:29]: for uniformity: `handleGetConsents` and `handleUpdateConsents`

File: plugins/flutter_aepconsent/ios/Classes/FlutterAEPConsentPlugin.m [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    [self handleUpdateConsent:call result:result];
+  } else {
+      result(FlutterMethodNotImplemented);
+  }
+}
+
+- (void)handleGetConsent:(FlutterMethodCall *) call result:(FlutterResult)result {
+     [AEPMobileEdgeConsent getConsents:^(NSDictionary* consents, NSError* error) {
+        result(consents);
+    }];
```
Comment (emdobrin) [inline plugins/flutter_aepconsent/ios/Classes/FlutterAEPConsentPlugin.m:40]: Are we just dropping the error here, is it not used/passed to dart?

File: plugins/flutter_aepconsent/ios/flutter_aepconsent.podspec [hunk 7]
```diff
@@ -0,0 +1,19 @@
+#
+# To learn more about a Podspec see http://guides.cocoapods.org/syntax/podspec.html
+#
```
Comment (emdobrin) [inline plugins/flutter_aepconsent/ios/flutter_aepconsent.podspec:3]: Let's remove

File: plugins/flutter_aepconsent/pubspec.yaml [hunk 8]
```diff
@@ -0,0 +1,37 @@
+name: flutter_aepconsent
```
Comment (emdobrin) [inline plugins/flutter_aepconsent/pubspec.yaml:1]: can we rename this to flutter_aepedgeconsent for uniformity with the other platforms? This applies for all the files/occurrences, except the dart class

File: plugins/flutter_aepconsent/pubspec.yaml [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+description: Official Adobe Experience Platform support for Flutter apps. AEPConsent is a new, innovative product from Adobe to help you easily validate SDK implementations.
```
Comment (emdobrin) [inline plugins/flutter_aepconsent/pubspec.yaml:3]: this seems inaccurate, maybe rephrase to something like
Official Adobe Experience Platform support for Flutter apps. The AEP Consent Collection plugin enables consent preferences collection from your Flutter app when using the [Adobe Experience Platform Mobile SDK](https://github.com/Adobe-Marketing-Cloud/acp-sdks) and the Edge Network extension.

File: plugins/flutter_aepconsent/android/src/main/java/com/adobe/marketing/mobile/flutter/FlutterAEPConsentPlugin.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+            handleGetConsent(result);
+    } else if("updateConsents".equals(call.method)) {
+            handleUpdateConsent(call.arguments);
+            result.success(null);
+    } else {
+      result.notImplemented();
+    }
+  }
+
+  private void handleGetConsent(final MethodChannel.Result result) {
```
Comment (emdobrin) [inline plugins/flutter_aepconsent/android/src/main/java/com/adobe/marketing/mobile/flutter/FlutterAEPConsentPlugin.java:63]: nit: rename to handleGetConsents, handleUpdateConsents

File: plugins/flutter_aepconsent/android/src/main/java/com/adobe/marketing/mobile/flutter/FlutterAEPConsentPlugin.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+          public void fail(final AdobeError adobeError) {
+              final String errorName = adobeError != null ? adobeError.getErrorName() : AdobeError.UNEXPECTED_ERROR.getErrorName();
+              result.error(Integer.toString(adobeError.getErrorCode()),"getConsents - Failed to retrieve consents",errorName);
+          }
+      });
+  }
+
+  private void handleUpdateConsent(final Object arguments) {
+        if (!(arguments instanceof Map)) {
+            Log.e(TAG, "Update Consent failed because arguments were invalid");
```
Comment (emdobrin) [inline plugins/flutter_aepconsent/android/src/main/java/com/adobe/marketing/mobile/flutter/FlutterAEPConsentPlugin.java:85]: ... were invalid, expected Map.

File: plugins/flutter_aepconsent/README.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
+# flutter_aepconsent
+
+[![pub package](https://img.shields.io/pub/v/flutter_aepassurance.svg)](https://pub.dartlang.org/packages/flutter_aepconsent) ![Build](https://github.com/adobe/aepsdk_flutter/workflows/Dart%20Unit%20Tests%20+%20Android%20Build%20+%20iOS%20Build/badge.svg) [![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
+
+`flutter_aepconsent` is a flutter plugin for the iOS and Android [AEPConsent SDK](https://aep-sdks.gitbook.io/docs/foundation-extensions/consent-for-edge-network) to allow for integration with Flutter applications. Functionality to enable the Assurance extension is provided entirely through Dart documented below.
```
Comment (emdobrin) [inline plugins/flutter_aepconsent/README.md:5]: Please update all occurrences of assurance in this file

File: plugins/flutter_aepconsent/lib/flutter_aepconsent.dart (hunk 1 above)
Comment (emdobrin) [reply plugins/flutter_aepconsent/lib/flutter_aepconsent.dart:25]: The key type should be string here. This error may be caused due to the usage inside the flutter sample app - https://github.com/adobe/aepsdk_flutter/pull/1/files#diff-72fd825ad17ce7800a2078e06f0dff1e12836ee59760465e78d9a5d222cebd15R228

File: plugins/flutter_aepconsent/lib/flutter_aepconsent.dart [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+      .invokeMethod<Map<dynamic, dynamic>>('getConsents')
+      .then((value) => value!);
+
+  /// Update the consent preferences stored in the Consent extension
```
Comment (emdobrin) [inline plugins/flutter_aepconsent/lib/flutter_aepconsent.dart:29]: Would be nice to have more detailed API docs in this public file, similar to the react native one

File: plugins/flutter_aepconsent/LICENSE [hunk 14]
```diff
@@ -0,0 +1,201 @@
+Apache License
```
[the remark this answers is not in the document]
Comment (cacheung) [reply plugins/flutter_aepconsent/LICENSE:1]: I will refer to the license in the root folder.  Let me discuss with @cdhoffmann about taking out the license from each plugin.

[the remark this answers is not in the document]
Comment (cdhoffmann) [reply plugins/flutter_aepconsent/LICENSE:1]: @emdobrin @cacheung The license is needed in each plugin directory because Flutter checks for a license in the plugin directory when you publish the plugin.

**Review by cdhoffmann — CHANGES_REQUESTED** [review]

File: example/pubspec.yaml [hunk 15]
```diff
@@ -22,6 +22,9 @@ dev_dependencies:
 
   flutter_aepassurance: ">= 1.0.0 <2.0.0"
 
+  flutter_aepconsent:
+    path: ../plugins/flutter_aepconsent
```
Comment (cdhoffmann) [inline example/pubspec.yaml:26]: Remember to refer to the actual plugin once it is published here. If i remember correctly i got errors when trying to use the local path from example app.

**Review by cdhoffmann — CHANGES_REQUESTED** [review]

File: plugins/flutter_aepedgeconsent/README.md [hunk 16]
```diff
@@ -0,0 +1,50 @@
+# flutter_aepedgeconsent
+
+[![pub package](https://img.shields.io/pub/v/flutter_aepassurance.svg)](https://pub.dartlang.org/packages/flutter_aepedgeconsent) ![Build](https://github.com/adobe/aepsdk_flutter/workflows/Dart%20Unit%20Tests%20+%20Android%20Build%20+%20iOS%20Build/badge.svg) [![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
+
+`flutter_aepedgeconsent` is a flutter plugin for the iOS and Android [AEPEdgeConsent SDK](https://aep-sdks.gitbook.io/docs/foundation-extensions/consent-for-edge-network) to allow for integration with Flutter applications. Functionality to enable the Assurance extension is provided entirely through Dart documented below.
```
Comment (emdobrin) [inline plugins/flutter_aepedgeconsent/README.md:5]: .. to enable the Consent extension is ..

File: example/lib/main.dart [hunk 17]
```diff
@@ -241,11 +284,11 @@ class _MyAppState extends State<MyApp> {
   Widget build(BuildContext context) {
     return MaterialApp(
         home: DefaultTabController(
-      length: 3,
+      length: 4,
       child: Scaffold(
         appBar: AppBar(
           bottom: TabBar(
-            tabs: [Text('Core'), Text('Identity'), Text('Assurance')],
+            tabs: [Text('Core'), Text('Identity'), Text('Assurance'), Text('Consent')],
```
[the remark this answers is not in the document]
Comment (cacheung) [reply example/lib/main.dart:284]: I plan to change the layout for sample app in another PR.

File: plugins/flutter_aepconsent/ios/Classes/FlutterAEPConsentPlugin.m (hunk 6 above)
Comment (cacheung) [reply plugins/flutter_aepconsent/ios/Classes/FlutterAEPConsentPlugin.m]: Working on this.

File: example/pubspec.yaml (hunk 15 above)
Comment (cacheung) [reply example/pubspec.yaml]: Will change it back to actual plugin when merge to staging.

File: plugins/flutter_aepconsent/lib/flutter_aepconsent.dart (hunk 1 above)
Comment (cacheung) [reply plugins/flutter_aepconsent/lib/flutter_aepconsent.dart]: Still looking into.

File: plugins/flutter_aepconsent/LICENSE (hunk 14 above)
[the remark this answers is not in the document]
Comment (cacheung) [reply plugins/flutter_aepconsent/LICENSE]: I am keeping the license in the consent directory.  Emilia points out we need to change the license to year 2022.  @cdhoffmann Do you want me to update the year for all the extensions license in this PR? Or want to do it later?

File: example/pubspec.yaml (hunk 15 above)
Comment (cacheung) [reply example/pubspec.yaml]: Updated back to actual plugin version.

File: plugins/flutter_aepconsent/README.md (hunk 12 above)
Comment (cacheung) [reply plugins/flutter_aepconsent/README.md]: Updated Readme.

File: plugins/flutter_aepconsent/ios/Classes/FlutterAEPConsentPlugin.m (hunk 6 above)
Comment (cacheung) [reply plugins/flutter_aepconsent/ios/Classes/FlutterAEPConsentPlugin.m]: Updated

File: plugins/flutter_aepconsent/lib/flutter_aepconsent.dart (hunk 1 above)
Comment (cacheung) [reply plugins/flutter_aepconsent/lib/flutter_aepconsent.dart]: Getting Map<object, object> response for getConsent.  Need to keep it to Map<String, dynamic> for this.

ReturnInDart

**Comment by cacheung** [comment]
Ready to be reviewed again.

File: plugins/flutter_aepedgeconsent/android/src/main/java/com/adobe/marketing/mobile/flutter/FlutterAEPEdgeConsentPlugin.java [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+                      result.success(consents);
+                  }
+              });
+          }
+
+          @Override
+          public void fail(final AdobeError adobeError) {
+              final String errorName = adobeError != null ? adobeError.getErrorName() : AdobeError.UNEXPECTED_ERROR.getErrorName();
+              result.error(Integer.toString(adobeError.getErrorCode()),"getConsents - Failed to retrieve consents",errorName);
+          }
```
Comment (emdobrin) [inline plugins/flutter_aepedgeconsent/android/src/main/java/com/adobe/marketing/mobile/flutter/FlutterAEPEdgeConsentPlugin.java:79]: L78 may still throw NPE if adobeError is null, so if you want to check for null case we should make this check before accessing it.
e.g.
```
final AdobeError error = adobeError != null ? adobeError : AdobeError.UNEXPECTED_ERROR;
result.error(Integer.toString(error.getErrorCode()),"getConsents - Failed to retrieve consents",error.getErrorName());
   }})..

File: plugins/flutter_aepedgeconsent/android/build.gradle [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+    }
+    lintOptions {
+        disable 'InvalidPackage'
+    }
+}
+
+dependencies {
+    api 'com.adobe.marketing.mobile:edgeconsent:1.+'
+    implementation 'com.android.support:appcompat-v7:28.0.0'
+    implementation 'com.android.support.constraint:constraint-layout:1.1.3'
```
Comment (emdobrin) [inline plugins/flutter_aepedgeconsent/android/build.gradle:39]: these dependencies are not required for this extension, let's remove

File: README.md [hunk 20]
```diff
@@ -8,6 +8,7 @@ This repository is a monorepo. It contains a collection of Adobe Experience Plat
 | ------------ | ------------------------------------------------------------ |
 | [Core](plugins/flutter_aepcore/README.md) (required)    | [![pub package](https://img.shields.io/pub/v/flutter_aepcore.svg)](https://pub.dartlang.org/packages/flutter_aepcore) |
 | [Assurance](plugins/flutter_aepassurance/README.md) | [![pub package](https://img.shields.io/pub/v/flutter_aepassurance.svg)](https://pub.dartlang.org/packages/flutter_aepassurance) |
+| [Consent](plugins/flutter_aepconsent/README.md) | [![pub package](https://img.shields.io/pub/v/flutter_aepconsent.svg)](https://pub.dartlang.org/packages/flutter_aepconsent) |
```
Comment (emdobrin) [inline README.md:11]: flutter_aepconsent -> flutter_aepedgeconsent

File: example/android/app/src/main/java/com/adobe/marketing/mobile/flutter/flutter_aepsdk_example/MyApplication.java [hunk 21]
```diff
@@ -5,11 +5,15 @@
 import android.util.Log;
 
 import com.adobe.marketing.mobile.*;
+//import com.adobe.marketing.mobile.edge.consent.Consent;
```
Comment (emdobrin) [inline example/android/app/src/main/java/com/adobe/marketing/mobile/flutter/flutter_aepsdk_example/MyApplication.java:8]: let's remove

File: plugins/flutter_aepedgeconsent/README.md [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+```
+
+## Usage
+### Consent
+
+For more detailed information on the Consent APIs, visit the documentation [here](https://aep-sdks.gitbook.io/docs/foundation-extensions/consent-for-edge-network)
+
+#### Registering the extension with AEPCore:
+
+ > Note: It is required to initialize the SDK via native code inside your AppDelegate and MainApplication for iOS and Android respectively.
```
Comment (emdobrin) [inline plugins/flutter_aepedgeconsent/README.md:30]: It is required to initialize the SDK via native code inside your AppDelegate (iOS) and MainApplication class (Android).

File: plugins/flutter_aepedgeconsent/README.md [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+// AppDelegate.m
+- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
+    [AEPMobileCore setWrapperType:AEPWrapperTypeFlutter];
+
+     // TODO: Set up the preferred Environment File ID from your mobile property configured in Data Collection UI
+    NSString* ENVIRONMENT_FILE_ID = @"YOUR-APP-ID";
+    
+    const UIApplicationState appState = application.applicationState;
+
+    NSArray *extensionsToRegister = @[AEPMobileIdentity.class,
```
Comment (emdobrin) [inline plugins/flutter_aepedgeconsent/README.md:57]: this should be AEPMobileEdgeIdentity

File: plugins/flutter_aepedgeconsent/README.md [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+  ...
+  // TODO: Set up the preferred Environment File ID from your mobile property configured in Data Collection UI
+  private final String ENVIRONMENT_FILE_ID = "YOUR-APP-ID";
+
+  @Override
+  public void on Create(){
+    super.onCreate();
+    ...
+    MobileCore.setApplication(this);
+    MobileCore.setLogLevel(LoggingMode.DEBUG);
```
Comment (emdobrin) [inline plugins/flutter_aepedgeconsent/README.md:89]: let's remove this from this sample

File: plugins/flutter_aepedgeconsent/README.md [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+        Identity.registerExtension();
+        Consent.registerExtension();
+        MobileCore.start(new AdobeCallback () {
+            @Override
+            public void call(Object o) {
+                MobileCore.configureWithAppID(ENVIRONMENT_FILE_ID);
+            }
+        });
+    } catch (InvalidInitException e) {
+        Log.e("MyApplication", String.format("Error while registering extensions %s", e.getLocalizedMessage()));
```
Comment (emdobrin) [inline plugins/flutter_aepedgeconsent/README.md:103]: this try/catch should not be required for these extensions

File: plugins/flutter_aepconsent/LICENSE (hunk 14 above)
[the remark this answers is not in the document]
Comment (cacheung) [reply plugins/flutter_aepconsent/LICENSE]: @cdhoffmann Should I change the Copyright year for Core extensions? Currently, they have year 2020.

**Review by emdobrin — APPROVED** [review]

File: plugins/flutter_aepedgeconsent/test/flutter_aepedgeconsent_test.dart [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+      ]);
+    });
+
+    test('returns correct result', () async {
+      expect(await Consent.extensionVersion, testVersion);
+    });
+  });
+}
+
+//TO DO: Add tests
```
Comment (cdhoffmann) [inline plugins/flutter_aepedgeconsent/test/flutter_aepedgeconsent_test.dart:49]: remove this

**Review by cdhoffmann — APPROVED** [review]
